add a base class for Matrix-free operators

davydden opened the pull request:
fixes https://github.com/dealii/dealii/issues/3242

This is mostly copy-paste from `matrix_free`  unit tests and organised into a stand-alone class to facilitate usage for an arbitrary operator to be defined in `apply()`.

@kronbichler ping.

@Rombur if you look at the base class, i would say there is enough code to grant its existence.

p.s. i might not be able to reply to feedback until Monday.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -30,6 +30,162 @@ DEAL_II_NAMESPACE_OPEN
+   * Base class for matrix-free operators which can be used both at the finest
+   * mesh or at a certain level in geometric multigrid.
+   */
+  template <int dim, int fe_degree, int n_q_points_1d = fe_degree+1, typename number = double>
```
Any reason why you put the templates `fe_degree` and `n_q_points_1d` on this class? You don't seem to use them. Even though there might be good reasons to use them in a derived class because `FEEvaluation` requires templates, I do not see why we would want to have them in the base class.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +361,260 @@ namespace MatrixFreeOperators
+  Base<dim,fe_degree,n_q_points_1d,number>::
+  initialize (const MatrixFree<dim,number> &data_,
+              const MGConstrainedDoFs &mg_constrained_dofs,
+              const MPI_Comm &mpi_communicator,
```
I think it would be better to remove `mpi_communicator` from this call (it is a bit strange to have a communicator here but not for the other method). The right thing to do would be to query the MPI communicator from the vector partitioner, `data_.get_vector_partitioner()->get_communicator()`, or alternatively, from the triangulation - but in that case you need to downcast to a parallel triangulation.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +361,260 @@ namespace MatrixFreeOperators
+                                                    const LinearAlgebra::distributed::Vector<number> &src) const
+  {
+    dst = 0;
+    vmult_add (dst,src);
```
How about adding an `apply_transpose` method that defaults to `apply_add` but a user can override in case a user wants to provide a different operator?

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -30,6 +30,162 @@ DEAL_II_NAMESPACE_OPEN
+     * Apply operator to @p src and write result in @p dst.
+     */
+    virtual void apply(LinearAlgebra::distributed::Vector<number> &dst,
+                       const LinearAlgebra::distributed::Vector<number> &src) const = 0;
```
I would call this method `apply_add` which is the assumption in `vmult_add`.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +372,299 @@ namespace MatrixFreeOperators
+        const_cast<LinearAlgebra::distributed::Vector<number>&>(src).local_element(edge_constrained_indices[i]) = 0.;
+      }
+
+    apply_add(dst,src);
```
@kronbichler : do we want to be able to have `Tapply_add` for `interface_up/down`? If so, i need to add `Tvmult_interface_down` and alike.

davydden replied:
done.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +372,299 @@ namespace MatrixFreeOperators
+
+  template <int dim, typename number>
+  void
+  Base<dim,number>::Tvmult_add (LinearAlgebra::distributed::Vector<number> &dst,
```
done.

kronbichler replied:
You use this because you would otherwise have to duplicate the manipulations in the vectors and the constraints, right?

davydden commented:
@kronbichler : I think I addressed your comments, please give it a second look. I hope this is general enough to be applicable to all those other cases which you do with your PhD students.  I used this class for Laplace (GMG and fine-scale) and mass matrix (fine-scale).

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +372,299 @@ namespace MatrixFreeOperators
+    if (!have_interface_matrices)
+      return;
+
+    LinearAlgebra::distributed::Vector<number> src_cpy (src);
```
No, one cannot get around this because one needs to zero out all but a few elements of the vector before going to the matrix-vector product. Thus, one needs a full vector of storage (and no need for the `const_cast` as compared to the usual vmult functions). It would be nice if one could merge this with the residual computation that precedes this call, but I have not yet found a neat way for doing that.

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -30,6 +30,224 @@ DEAL_II_NAMESPACE_OPEN
+                   const LinearAlgebra::distributed::Vector<number> &src,
+                   const bool transpose) const;
+
+
```
Remove one empty line

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -30,6 +30,224 @@ DEAL_II_NAMESPACE_OPEN
+                           const bool transpose) const;
+  };
+
+
```
add one more empty line

kronbichler commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +423,350 @@ namespace MatrixFreeOperators
+  {
+  }
+
+
```
Please use three empty lines consistently between the function definitions.

kronbichler commented:
> I hope this is general enough to be applicable to all those other cases which you do with your PhD students.

We'll have a look at this when this gets merged. It definitely will help a number of cases.

davydden commented:
> You use this because you would otherwise have to duplicate the manipulations in the vectors and the constraints, right?

if you mean those `mult_add` and alike -- yes, this is to avoid a lot of code duplication where the difference is only `apply_add` vs `Tapply_add`.

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -30,6 +30,224 @@ DEAL_II_NAMESPACE_OPEN
+    /**
+     * Compute diagonal of this operator. To be implemented in derived classes.
+     */
+    virtual void compute_diagonal () = 0;
```
I would keep it as pure virtual which would make users to at least add a dummy implementation with `AssertThrow(false, ExcNotImplemented())`.

davydden commented:
I updated the PR.

kronbichler commented:
Regarding the transpose version of the interface matrix-vector products: The multigrid algorithms for continuous elements do not need the transpose. The interface is a bit confusing because the `MGInterfaceMatrix` we use applies the `Tvmult` operation on the `vmult_interface_up` but if you check in the paper by Janssen & Kanschat (SISC 2011, sec. 2.5) you see that actually only the non-transpose version is necessary. Since the only application I see for the interface multiplications is in the multigrid residual context, I would not add the transpose version of those two methods to keep the interface to what we think is necessary.

kronbichler commented:
The other changes look nice! :+1:

davydden commented:
>  Since the only application I see for the interface multiplications is in the multigrid residual context, I would not add the transpose version of those two methods to keep the interface to what we think is necessary.

ok, will do.

kronbichler commented:
Please squash the commits as appropriate and we should be good to go.

kronbichler commented:
I'd say we wait another day with merging so @Rombur also gets a chance to comment this code.

Rombur commented on tests/matrix_free/parallel_multigrid_adaptive_02.cc:
```diff
@@ -0,0 +1,453 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2014 - 2016 by the deal.II authors
```
Since it's a new test, you only need 2016.

davydden replied:
good catch, i often see this in other's PR, but still have it mine :smile:

davydden commented on include/deal.II/matrix_free/operators.h:
```diff
@@ -205,6 +397,322 @@ namespace MatrixFreeOperators
+  void
+  Base<dim,number>::clear ()
+  {
+    data = NULL;
```
not sure about it, from documentation of `SmartPointer::clear()`:

> Delete the object pointed to and set the pointer to zero.

I certainly don't want to delete the object as it can be used elsewhere.

kronbichler replied:
SmartPointer does not manage any memory per default as its main purpose is to keep track of the uses of the object pointed to. So calling `clear()` is clearly wrong. However, assigning a new pointer is exactly right, as it calls `unsubscribe` on the old owner and subscribes to the new one in case it is not null. So I'd say it looks perfectly fine here.

davydden commented:
@Rombur @kronbichler :  I presume this can be merged then?

Rombur commented:
Yes, it's good for me.
